tensorflow-keras/src/main/java/org/tensorflow/keras/backend/K.java

```diff
@@ -0,0 +1,713 @@
+            tf.constant(new long[] {0}),
+            tf.constant(new long[] {inputShape.numDimensions() - 1}));
+    cost = tf.reshape(cost, outputShape);
+    if (tf.scope().env().isGraph() && !shape.hasUnknownDimension()) {
+      long[] array = shape.asArray();
+      long[] newArray = new long[array.length - 1];
+      if (axis < 0) {
+        axis = shape.numDimensions() + axis;
+      }
+      for (int i = 0; i < axis; i++) {
+        newArray[i] = shape.size(i);
+      }
+      for (int i = axis + 1; i < shape.numDimensions(); i++) {
+        newArray[i - 1] = shape.size(i);
+      }
+      Shape newShape = Shape.of(newArray);
+      cost = tf.reshape(cost, tf.constant(newShape.asArray()));
+    }
+
+    if (convertToFloat32) {
+      cost = tf.dtypes.cast(cost, logits.asOutput().dataType());
+    }
+    return cost;
+  }
+
+  public static <T extends TType> Operand<T> map(
+      Operand<T> input, Function<Operand<T>, Operand<T>> mapFunc) {
+    return null;
+  }
+
+  public static long[] concatenate(long first, long... remaining) {
+    long[] dims = new long[remaining.length + 1];
+    System.arraycopy(remaining, 0, dims, 1, remaining.length);
+    dims[0] = first;
+    return dims;
+  }
+
+  private static Map<ExecutionEnvironment, Map<String, Integer>> uidMap = new HashMap<>();
+
+  /**
+   * Associates a string prefix with an integer counter in a TensorFlow graph.
+   *
+   * <p>Example:
+   *
+   * <pre>
+   * get_uid('dense')
+   * 1
+   * get_uid('dense')
+   * 2
+   * </pre>
+   *
+   * @param tf the TensorFlow Ops
+   * @param prefix String prefix to index.
+   * @return Unique integer ID.
+   */
+  public static int getUid(Ops tf, String prefix) {
+    ExecutionEnvironment env = tf.scope().env();
+    Map<String, Integer> uids = uidMap.get(env);
+    if (uids == null) {
+      uids = new HashMap<>();
+      uidMap.put(env, uids);
+    }
+    Integer id = uids.get(prefix);
+    if (id == null) {
+      id = 0;
+    } else {
+      id++;
+    }
+
+    uids.put(prefix, id);
+    return id;
+  }
+
+  /**
+   * returns the larger DataType between the two.
+   *
+   * @param a the first DataType to compare
+   * @param b the second DataType to compare
+   * @return the wider DataType
+   */
+  public DataType wider(DataType a, DataType b) {
+    return a.byteSize() < b.byteSize() ? b : a;
+  }
+
+  /**
+   * returns the smaller DataType between the two.
+   *
+   * @param a the first DataType to compare
+   * @param b the second DataType to compare
+   * @return the smaller DataType
+   */
+  public DataType narrower(DataType a, DataType b) {
+    return a.byteSize() > b.byteSize() ? b : a;
+  }
+
+  public <T extends TNumber> NdArraySequence getTensorValue(Ops tf, Operand<T> operand) {
+    DataType dtype = operand.asOutput().dataType();
+    if (tf.scope().env().isGraph()) {
+      try (Session session = new Session((Graph) tf.scope().env())) {
+        if (dtype.equals(TInt32.DTYPE)) {
+          try (Tensor<TInt32> result =
```

Comment:
To make this usable, `NdArraySequence` must be typed. If we decide to go ahead with [this PR](https://github.com/tensorflow/java/pull/92), it could make things easier but I'll need to double-check before confirming this. Otherwise, we will need one method per type.